### PR TITLE
fix: reduce some false positive matches

### LINKED ISSUE #21: Semantic version tests are not being passed
The [official semver site](https://semver.org/) has its own [regex listings and tests](https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string).

The following link shows how this regex compares against the semver tests: https://regex101.com/r/3SKyp2/1

### PR DESCRIPTION
**Pull request by antongolub**
* fix: handle some false positive matches
~~* feat: introduce SEMVER_REGEX const~~
* test: check what exactly is captured by regex
* test: verify test cases via semver
* chore: up deps, apply xo

closes #21

### CODE REVIEW

**Comment by antongolub** [comment]
Reasonable. https://devina.io/redos-checker, the results are similar: 

Before:
image

After:
image

**Comment by antongolub** [comment]
@sindresorhus, can you trigger tests again?

File: index.js [hunk 1]
```diff
@@ -1,3 +1,5 @@
 export default function semverRegex() {
-	return /(?:(?<=^v?|\sv?)(?:(?:0|[1-9]\d{0,9}?)\.){2}(?:0|[1-9]\d{0,9}?)(?:-(?:0|[1-9]\d*?|[\da-z-]*?[a-z-][\da-z-]*?){0,100}?(?:\.(?:0|[1-9]\d*?|[\da-z-]*?[a-z-][\da-z-]*?))*?){0,100}?(?:\+[\da-z-]+?(?:\.[\da-z-]+?)*?){0,100}?\b){1,200}?/gi;
+	return /(?<=^v?|\sv?)(?:(?:0|[1-9]\d{0,9}?)\.){2}(?:0|[1-9]\d{0,9})(?:-(?:0|[1-9]\d*|[\da-z-]*?[a-z-][\da-z-]*?)){0,100}?(?=$| |\+|\.)(?:(?<=-\S+)(?:\.(?:[\da-z-]*?[a-z-][\da-z-]*|0|[1-9]\d*)){1,100}?)?(?!\.)(?:\+(?:[\da-z-]+(?:\.[\da-z-]+)*){1,100}?(?!\w))?(?!\+)/gi;
 }
+
+export const SEMVER_REGEX = semverRegex();
```
Comment (sindresorhus) [inline index.js:5]: There's a good reason the export is a function. Regexes in JS are mutable if they use the `g` flag.

Comment (antongolub) [reply index.js:5]: Object.freeze?

Comment (sindresorhus) [reply index.js:5]: Not sure that will work, but `Object.freeze` has a big performance penalty. I don't really see the point of having two exports then when the second one has very little benefit.

**Comment by antongolub** [comment]
Mb add named `semverRegex` export too?

**Comment by antongolub** [comment]
I have to admit, that I have no idea how to implement trully redos-safe checker: we have a fairly wide range of chunk combinations and the semver spec does not impose their length restrictions.

**Comment by antongolub** [comment]
@sindresorhus,

Anyway, is any chance to land this?

**Comment by sindresorhus** [comment]
> we have a fairly wide range of chunk combinations and the semver spec does not impose their length restrictions.

But we can restrict them to some realistic values.

**Comment by sindresorhus** [comment]
> Anyway, is any chance to land this?

To be honest, I don't know where to start reviewing. The regex is just way too complicated now. Maybe if you could split up the regex into manageable parts. Like the main version component, prerelease, and build.

If you need it right away, I would suggest just doing:

```sh
npm install 'antongolub/semver-regex#fix-false-positive-matches'
```

**Comment by antongolub** [comment]
We have dozens packages depending on semver-regex, so I'd like to provide the fix here. Maybe we just need to extnd redos tests. And it seems that if the new regex completely covers the spec, perhaps this will be its last change)

**Comment by antongolub** [comment]
Leaving things as they are seems wrong: 
image
This misbehavior has been here for so long time and probably has a bunch of side effects, that the fix will have to be published as 5.0.0.

**Comment by antongolub** [comment]
@sindresorhus,

image

**Comment by antongolub** [comment]
> npm install 'antongolub/semver-regex#fix-false-positive-matches'

@sindresorhus, let's publish this as beta, so we could override the prev version via yarn resolutions.

**Comment by sindresorhus** [comment]
CI is failing

**Comment by antongolub** [comment]
So close)
```
Execution time: 20
t.true(difference < 20, `Execution time: ${difference}`);
```

**Comment by antongolub** [comment]
✖ invalid version does not cause catatrophic backtracking Execution time: 26
t.true(difference < 25, `Execution time: ${difference}`);

**Comment by antongolub** [comment]
I'll set to 50)

**Comment by antongolub** [comment]
@sindresorhus, your turn

**Comment by sindresorhus** [comment]
> Mb add named semverRegex export instead? We'd like to avoid default exports in our codebase.

I prefer using default export when it makes sense and it makes no sense to export both a default and named export with the same thing.

**Comment by sindresorhus** [comment]
Thanks :)

**Comment by sindresorhus** [comment]
https://github.com/sindresorhus/semver-regex/releases/tag/v4.0.4

**Pull request merged**